### ISSUE TITLE
Training stops after step 4000

### ISSUE THREAD

**Issue by FestinaI** [POSTECH-CVLab/PyTorch-StudioGAN#136]
Training always stops after this step:

Visualize (num_rows x 8) fake image canvans.
Save image canvas to output/figures/CUSTOM-ReACGAN-train-2022_03_23_14_40_18/generated_canvas_4000.png
Start Evaluation (4000 Step): CUSTOM-ReACGAN-train-2022_03_23_14_40_18 generate images and stack features (36808 images).

I'm training ReACGAN with my own dataset. However, the training stops always after this step and there are no logs to know why!

Any help or idea would be highly appreciated!

**Comment by MINGUKKANG** [comment]
Did you use the -metrics prdc option?

It takes much time (in my case, 6 hours for ImageNet) to calculate the prdc. 

So it is highly recommended for you to turn on is and fid options only for fast training.

Thank you.

Best,

Minguk

**Comment by FestinaI** [comment]
Thank you for your reply! I was using the prdc metric as well, I'm trying now without and will see. 

There is another thing that I'm not sure if it's correct:
When loading the train and validation data, based on the logs it says that the length of eval dataset is the same as train dataset. 

The logs: 
Load CUSTOM train dataset.
Train dataset size: 36808

then for validation: (takes the same path as the training set)
Load CUSTOM train dataset.
Eval dataset size: 36808

The way I'm training is: "python3 src/main.py -t -metrics fid -cfg src/configs/CIFAR10/ReACGAN.yaml -data data/CUSTOM -save output/"

I added the folders: data/CUSTOM/train and data/CUSTOM/valid.  - as described in the documentation. 


Any ideas are highly appreciated!
Thank you

**Comment by FestinaI** [comment]
Hi thank you for your feedback!
I changed only the DATA part inside yaml file. Also, for the MODEL i changed the z_dim only. I guess the other parts should remain as is right? 

As for the interrupted training, this is still happening. 
<img width="1043" alt="Screenshot 2022-03-28 at 10 11 24" src="https://user-images.githubusercontent.com/12886033/160355055-806270a6-132b-4916-84fa-ce24ef38c9ad.png">


This is my command: CUDA_VISIBLE_DEVICES=0,1,2,3 python3 src/main.py -t -metrics fid -cfg src/configs/CIFAR10/ReACGAN.yaml -data data/CUSTOM -save output/ >> logs.txt &

I assume my dataset is not that large  in the end. Number of samples are: 36808, with 7 classes. 

Any ideas are highly appreciated!

**Comment by MINGUKKANG** [comment]
<img width="1033" alt="image" src="https://user-images.githubusercontent.com/30516904/160359680-446d65f9-ae0b-46d9-9af1-73d9ea972c82.png">

That procedure is conducted in "generate_images_and_stack_features" function, so could you find the exact point where the training stops?

You can use the \textit{pdb} debugger to identify this.

Thank you.

**Comment by FestinaI** [comment]
Thank you for your feeback! I'm training now not as a background process and it didn't stop yet. 
However, the training is very slow - 20% for 9 hours now in 4 GPUs. I'm assuming bcs the training data is being used for evaluation?

Do you have any tips on test data? 
I'm using my own custom dataset (medical data). Is it recommended to have a test folder inside validation folder?  I didn't see it in the documentation. 

Also, is there any way to resume training from the last checkpoint? Or should I include this functionality myself?

Thank you!
I appreciate any feedback!

**Comment by MINGUKKANG** [comment]
However, the training is very slow - 20% for 9 hours now in 4 GPUs. I'm assuming bcs the training data is being used for evaluation?

=> I am not sure whether training dataset used for evaluation causes the slow training. If you use -ref "valid" or "test" option, you can measure the metrics not using "training" dataset but "valid" or "test" dataset.

Do you have any tips on test data?
I'm using my own custom dataset (medical data). Is it recommended to have a test folder inside validation folder? I didn't see it in the documentation.

=> If you follow the data structure I wrote in README, you can change the dataset for evaluation using -ref "train", -ref "valid", or -ref "test" option.

Also, is there any way to resume training from the last checkpoint? Or should I include this functionality myself?

=> Yes, you can resume your training using -ckpt CHECKPOINT_PATH option.
Please refer to README file.

Best,

Minguk

**Comment by FestinaI** [comment]
Thanks a lot for this quick feedback! 
Do you recommend to use -ref "valid" or "train" ?

**Comment by MINGUKKANG** [comment]
Yes, I recommend you to use -ref "valid" during training. 

After training is done, you can evaluate your trained model using -ref "test" option.

Thk.

**Comment by FestinaI** [comment]
Thank you!
I also need to ask regarding the image generation process after the training is finished. 
Which command should I use to only generate images in the end and not train? How do I specify the seeds for the generator? I want to generate let's say 8000 samples for class 2.

**Comment by MINGUKKANG** [comment]
How do I specify the seeds for the generator? 
==> --seed SEED_NUMBER 

I want to generate let's say 8000 samples for class 2.
==> You can use the command below to generate fake images from a pre-trained generator.
```
~ options -ckpt CKPT_PATH_FOR_GAN -sf --seed 1234 -sf_num 8000*num_classes
```

The above command will generate approximately 8000 images per class.

Thank you.

**Comment by FestinaI** [comment]
The sf_num is not recognizable on the master branch!! Are there any known issues?

**Comment by FestinaI** [comment]
I saw in one of the open issues before mine. 
<img width="978" alt="Screenshot 2022-03-31 at 10 40 52" src="https://user-images.githubusercontent.com/12886033/161014290-f6c75833-c5c2-43f3-8672-ed6293c2a8ea.png">


Is this change in the master branch? 
This command doesn't work: CUDA_VISIBLE_DEVICES=0 python3 src/main.py -cfg CONFIG_PATH -save SAVE_DIR -ckpt CKPT_PATH -sf -sf_num NUMBER_OF_IMAGES_TO_GENERATE -metrics none 

The branch mentioned in the screenshot above doesn't exist also. 

Any help appreciated!

**Comment by FestinaI** [comment]
I pulled the last changes from master (for the reason to generate images) and while training with this command as before (used to work):

CUDA_VISIBLE_DEVICES=1,2,3,4 python3 src/main.py -t -metrics fid -ckpt=output/checkpoints/CUSTOM-ReACGAN-train-2022_03_29_19_15_17/ -ref "valid" -cfg src/configs/CIFAR10/ReACGAN.yaml -data data/CUSTOM -save output/ 

the ref "valid" attribute is not known anymore
<img width="720" alt="Screenshot 2022-03-31 at 11 05 18" src="https://user-images.githubusercontent.com/12886033/161019387-4184df0b-0717-42d5-9f1d-599d477adee9.png">


Do i have to make any changes?

Thank you!

**Comment by FestinaI** [comment]
Also just to add some comments when generating data. 

This is the command I used to generate data with the last updates from master:

python3 src/main.py -cfg src/configs/CIFAR10/ReACGAN.yaml -save results/ -ckpt=output/checkpoints/CUSTOM-ReACGAN-train-2022_03_29_19_15_17/ -sf --seed 1234 -sf_num 70 -data data/CUSTOM -ref "valid"

I had to add -ref "valid" and also -data, I'm not sure why is this needed. Also, the sf_num it doesn't distribute the generated samples like 10 for each class (in my case I have 7 classes).

**Comment by MINGUKKANG** [comment]
Also, is there a way to generate data with different checkpoints from different steps?
=> No, StudioGAN does not support this function.

the ref "valid" attribute is not known anymore
=> plase change self.num_eval[self.RUN.ref_dataset] to self.num_eval["test"].

Best,

Minguk

**Comment by MINGUKKANG** [comment]
please print(self.num_eval) 

and input an appropriate key for evaluation.

Thank you.
